File: insert-delete-getrandom-o1/insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    unordered_map<int, int>u;
    vector<int>v;
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(u.find(val)!=u.end()) return false;
        u[val] = v.size();
        v.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if(u.find(val)==u.end()) return false;
        int idx = u[val];
        if(idx<v.size()-1) {
            swap(v[idx], v[v.size()-1]);
            u[v[idx]] = idx;
        }
        u.erase(val);
        v.pop_back();
        return true;
    }
    
    int getRandom() {
        return v[rand()%v.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.dict = {}
        self.list = []
        

    def insert(self, val: int) -> bool:
        if val not in self.dict:
            self.dict[val] = len(self.list)
            self.list.append(val)
            return True
        return False
        
    def remove(self, val: int) -> bool:
        if val not in self.dict:
            return False
        index = self.dict[val]
        last = self.list[-1]
        self.list[index] = last
        self.dict[last] = index
        self.list.pop()
        del self.dict[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.list)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()